File: Src/Icm.ContextConsole/Context/BaseContext.vb
Imports Icm.Localization

''' <summary>
''' A base context implements many of the methods of IContext based on 4 of them.
''' </summary>
''' <remarks></remarks>
Public MustInherit Class BaseContext
    Implements IContext


    Private ReadOnly _actionFinder As IActionFinder
    Protected locRepo As ILocalizationRepository
    Protected extLocRepo As ILocalizationRepository

    Property Interactor As IInteractor

    Public Sub New()
        _actionFinder = New ReflectionActionFinder
    End Sub

    Public Sub New(actionFinder As IActionFinder)
        _actionFinder = actionFinder
    End Sub


    Protected Overridable Sub BaseInitialize(app As IApplication) Implements IContext.Initialize
        locRepo = app.InternalLocRepo
        extLocRepo = app.ExternalLocRepo
        Interactor = app.Interactor
        Initialize()
    End Sub

    Protected Overridable Sub Initialize()
    End Sub

    Public Function Name() As String Implements IContext.Name
        Dim ctlTypeName = Me.GetType.Name
        Return ctlTypeName.Substring(0, ctlTypeName.Length - "Context".Length).ToLower
    End Function


    Public Function Synonyms() As IEnumerable(Of String) Implements IContext.Synonyms
        Dim synAttr = Me.GetType().GetAttribute(Of SynonymAttribute)(True)
        Dim result As New List(Of String)
        If synAttr IsNot Nothing Then
            result.AddRange(synAttr.Synonyms.Select(Function(str) str.ToLower))
        End If
        Return result
    End Function

    Public Function Named(contextName As String) As Boolean
        Return Name() = contextName OrElse Synonyms().Contains(contextName)
    End Function

    Public ReadOnly Property ActionFinder As IActionFinder Implements IContext.ActionFinder
        Get
            Return _actionFinder
        End Get
    End Property
End Class

File: Src/Icm.ContextConsole/Context/BaseContext.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Linq;
using Icm.Localization;
using Icm.Reflection;

/// <summary>
/// A base context implements many of the methods of IContext based on 4 of them.
/// </summary>
/// <remarks></remarks>
public abstract class BaseContext : IContext
{


	private readonly IActionFinder _actionFinder;
	protected ILocalizationRepository locRepo;

	protected ILocalizationRepository extLocRepo;
	public IInteractor Interactor { get; set; }

	public BaseContext()
	{
		_actionFinder = new ReflectionActionFinder();
	}

	public BaseContext(IActionFinder actionFinder)
	{
		_actionFinder = actionFinder;
	}


	protected virtual void BaseInitialize(IApplication app)
	{
		locRepo = app.InternalLocRepo;
		extLocRepo = app.ExternalLocRepo;
		Interactor = app.Interactor;
		Initialize();
	}
	void IContext.Initialize(IApplication app)
	{
		BaseInitialize(app);
	}

	protected virtual void Initialize()
	{
	}

	public virtual string Name()
	{
		var ctlTypeName = this.GetType().Name;
		return ctlTypeName.Substring(0, ctlTypeName.Length - "Context".Length).ToLower();
	}


	public virtual IEnumerable<string> Synonyms()
	{
		var synAttr = this.GetType().GetAttribute<SynonymAttribute>(true);
		List<string> result = new List<string>();
		if (synAttr != null) {
			result.AddRange(synAttr.Synonyms.Select(str => str.ToLower()));
		}
		return result;
	}

	public bool Named(string contextName)
	{
		return Name() == contextName || Synonyms().Contains(contextName);
	}

	public IActionFinder ActionFinder {
		get { return _actionFinder; }
	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
